### DISCUSSION TITLE
Incompatible QoS on topic `/battery_state`

### DISCUSSION THREAD

**Discussion by chrisbuerginrogers** [iRobotEducation/create3_docs#23] [Questions] [answered]
So I am trying to write a very simple subscriber - but get the error [WARN] [1643999034.968400508] [minimal_subscriber]: New publisher discovered on topic '/battery_state', offering incompatible QoS. No messages will be received from it. Last incompatible policy: RELIABILITY
 when I try to "spin it up" - do I have to set the QoS level in the creation of the subscription?

import sys,rclpy
from rclpy.node import Node
from sensor_msgs.msg import BatteryState
from irobot_create_msgs.msg import Button 

class MinimalSubscriber(Node):

    def __init__(self):
        super().__init__('minimal_subscriber')
        self.subscription = self.create_subscription(
            BatteryState,'/battery_state',self.listener_callback,10)
        self.subscription  # prevent unused variable warning

    def listener_callback(self, msg):
        self.get_logger().info('I heard: "%s"' % msg.data)

def main(args=None):
    rclpy.init(args=args)

    minimal_subscriber = MinimalSubscriber()
    try: 
        rclpy.spin(minimal_subscriber)
    except KeyboardInterrupt:
        print('Caught keyboard interrupt')
    except BaseException:
        print('Exception:', file=sys.stderr)
    finally:
        minimal_subscriber.destroy_node()
        rclpy.shutdown()

**Comment by justinIRBT** [comment, marked as answer]
An answer to this question is explained in more detail here:
https://answers.ros.org/question/332207/cant-receive-data-in-python-node/

The short answer is instead of using `10` in the `create_subscription` call, use:
```
from rclpy.qos import qos_profile_sensor_data

self.subscription = self.create_subscription(
        BatteryState,'/battery_state',self.listener_callback,qos_profile_sensor_data)
```

**Reply by chrisbuerginrogers** [reply]
Thanks - that definitely got me farther - but it still does not receive anything - instead I get an Exception (with nothing returned from sys.stderr).  Any idea why?  thanks

**Reply by justinIRBT** [reply]
The problem is in your listener_callback logger printout:
`self.get_logger().info('I heard: "%s"' % msg.data)`
BatteryState doesn't have a data field.  If you just print the msg, it works for me
`self.get_logger().info('I heard: "%s"' % msg)`

**Reply by chrisbuerginrogers** [reply]
Oh shoot - right - that worked great - sorry - too focussed on other parts of the code...
